### ISSUE TITLE
Extremely poor data loader performance, taking too much time to load first batch.

### ISSUE THREAD

**Issue by braindotai** [tensorflow/tensorflow#47687]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.5 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Not applicable.
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.4.1
- Python version: 3.7.10
- CUDA/cuDNN version: 11.0
- GPU model and memory: Tesla T4, 15109MiB.

**Describe the current behaviour**
I'm trying to create a custom data loader for training data, where the data has been downloaded from `tensorflow_datasets`,
and a number of data preprocessing have been applied.

But the data loader is taking too much time to load the first batch of each epoch. 

```python
for lrs, hrs in train_data:
    break
```

The code above is taking well over 5 minutes, which means it is taking 5 minutes just to load the first batch. **Though, once the first batch is loaded, then the loading of next batches are faster, and the rest of the batches are being loaded without any issues, each taking not more than 1 second**

**But again once the loop/epoch is over, then starting through the first batch again is taking 5 minutes.**

I've tested the issue with/without `model.fit` function, facing the same issue in both cases.

**Describe the expected behavior**

The loading of the first batch is actually expected to be slow, but not this slow, cuz even the training of the model each batch is not taking this much time.

**Standalone code to reproduce the issue**
Below is the completed code to reproduce the issue.

```python
HR_SIZE = 128
SCALE = 4
BATCH_SIZE = 32

train_data = tfds.load(f'div2k/bicubic_x4', split = 'train')
# this line will download about 4GB of data....
# so using a colab notebook is recommended to reproduce the issue.

@tf.function
def random_crop(example):
    lr = example['lr']
    hr = example['hr']

    lr_crop_size = HR_SIZE // SCALE
    lr_shape = tf.shape(lr)[:2]

    lr_w = tf.random.uniform(shape = (), maxval = lr_shape[1] - lr_crop_size + 1, dtype = tf.int32)
    lr_h = tf.random.uniform(shape = (), maxval = lr_shape[0] - lr_crop_size + 1, dtype = tf.int32)

    hr_w = lr_w * SCALE
    hr_h = lr_h * SCALE

    lr_cropped = lr[lr_h:lr_h + lr_crop_size, lr_w: lr_w + lr_crop_size]
    hr_cropped = hr[hr_h:hr_h + HR_SIZE, hr_w: hr_w + HR_SIZE]

    return lr_cropped, hr_cropped

@tf.function
def random_horizontal_flip(lr, hr):
    return tf.cond(tf.random.uniform(shape = (), maxval = 1) < 0.5,
                   lambda: (lr, hr),
                   lambda: (tf.image.flip_left_right(lr),
                            tf.image.flip_left_right(hr)))

@tf.function
def random_vertical_flip(lr, hr):
    return tf.cond(tf.random.uniform(shape = (), maxval = 1) < 0.5,
                   lambda: (lr, hr),
                   lambda: (tf.image.flip_up_down(lr),
                            tf.image.flip_up_down(hr)))

@tf.function
def random_gain(lr, hr):
    gain_delta = tf.random.uniform(shape = (), minval = 0.7, maxval = 1.6)
    return tf.cond(tf.random.uniform(shape = (), maxval = 1) < 0.5,
                   lambda: (lr, hr),
                   lambda: (tf.image.adjust_gamma(lr, gamma = 1.0, gain = gain_delta),
                            tf.image.adjust_gamma(hr, gamma = 1.0, gain = gain_delta)))

@tf.function
def random_contrast(lr, hr):
    contrast_delta = tf.random.uniform(shape = (), minval = 0.8, maxval = 1.2)
    return tf.cond(tf.random.uniform(shape = (), maxval = 1) < 0.5,
                   lambda: (lr, hr),
                   lambda: (tf.image.adjust_contrast(lr, contrast_factor = contrast_delta),
                            tf.image.adjust_contrast(hr, contrast_factor = contrast_delta)))

@tf.function
def random_rotate(lr, hr):
    rn = tf.random.uniform(shape = (), maxval = 4, dtype=tf.int32)
    return tf.image.rot90(lr, rn), tf.image.rot90(hr, rn)

train_data = train_data.map(random_crop, num_parallel_calls = tf.data.AUTOTUNE)
train_data = train_data.map(random_horizontal_flip, num_parallel_calls = tf.data.AUTOTUNE)
train_data = train_data.map(random_vertical_flip, num_parallel_calls = tf.data.AUTOTUNE)
train_data = train_data.map(random_gain, num_parallel_calls = tf.data.AUTOTUNE)
train_data = train_data.map(random_contrast, num_parallel_calls = tf.data.AUTOTUNE)
train_data = train_data.map(random_rotate, num_parallel_calls = tf.data.AUTOTUNE)

train_data = train_data.batch(BATCH_SIZE, drop_remainder = True)
train_data = train_data.prefetch(tf.data.AUTOTUNE)
train_data = train_data.shuffle(800)

for lrs, hrs in train_data:
    break
```

And I'm running all of my code in google colab.

**Comment by Saduf2019** [comment]
@braindotai 
The code shared is not reproducible, please find [gist here](https://colab.research.google.com/gist/Saduf2019/5e00a3e32ff530b8a9c826ee024d1980/untitled561.ipynb), and share the colab gist with the error reporeted.

**Comment by braindotai** [comment]
Here's [the link](https://colab.research.google.com/gist/braindotai/3b917caf1d9f72569e5af1162d77ca42/untitled561.ipynb#scrollTo=cRyFuojRqV6f) to the notebook for reproducing the issue.

**Comment by braindotai** [comment]
@AdityaKane2001 
So does that mean this is expected behaviour?
If yes, then is there any way to apply the `.map` functions batch by batch instead of blocking the main thread process?

I'm really worried because my data is only about 3 GB... with only 1600 quality images, and still in my case just the preprocessing of data is taking well over 4 to 5 minutes... which is surprisingly way greater than training my model for 1 epoch.

So in the case of a much larger and heavy dataset, using `tf.data` would not be a good experience... cuz in that case we'll just be spending most of our time applying `.map` functions. 😐

**Comment by AdityaKane2001** [comment]
@braindotai 
I logged the shapes of each of the images in the dataset. The shapes of the images are not the same, some images have the same shape whereas others don't. This causes retracing of all the mapping functions, and that may be the reason of poor performance. If your application permits, please try adding a resizing function. Also, make sure that your mapping functions are taking  tf constructs as input. Else, the graphs will be retraced on every call.

**Comment by braindotai** [comment]
> Also, make sure that your mapping functions are taking tf constructs as input. Else, the graphs will be retraced on every call.

I'm not sure if I know how to implement this. Could you please give an example for this?

**Comment by AdityaKane2001** [comment]
@braindotai 
Add a `print('<function name>')` statement in all functions. When the dataset is transformed for the first time, these statements will appear each time a function is called. If it appears exactly once, then it's alright, as all functions decorated with `@tf.function` must be traced atleast once.
Make sure that all arguments of a given function are TensorFlow constructs only, which means that arguments must be one of the following:
`Tensor`,`TensorSpec`,`TensorArray`,etc..
If they are not, then your function is getting retraced multiple times, adding to the time and RAM usage.

**Comment by braindotai** [comment]
@AdityaKane2001 
Thanks for the support, I'll try it out and get back to you.

**Comment by kdonbekci** [comment]
Hey @braindotai  and @AdityaKane2001 , just wanted to also point out that in the code provided, a very costly `shuffle` operation is being done. 

I checked how large the original data is with:
```
train_data = tfds.load(f'div2k/bicubic_x4', split = 'train')
print(train_data.cardinality()) # --> 800
```

The notebook in question has, as a final operation, `train_data = train_data.shuffle(800)`. Note that the dataset has already been batched into 32 at this point. With this shuffle op, `tf.data` actually iterates AND PREPROCESSES the entire dataset a whopping 32 times unnecessarily before producing the 1st batch. The correct `shuffle_buffer_size` to achieve a perfect shuffle should be 800/32 = 25 at this stage in the pipeline.

Furthermore, I would suggest pushing the shuffle earlier into the `tf.data` pipeline as the inputs do not need to be preprocessed before they are shuffled. In fact the `tfds.load` function has a parameter for shuffling files called `shuffle_files` which you should probably use. (especially if you are working on Colab with slow IO)

As a final suggestion, similar to the comments above regarding retracing and input tensors with different sizes (which cause retracing), I think you might be better off by undecorating your functions with `tf.function`. 

here are some benchmarks:

1. Moving `.shuffle` to beginning of pipeline. --> 64 seconds for 1 COMPLETE ITERATION OVER DATASET
2. Removing `.shuffle` altogether. --> 61 seconds for 1 COMPLETE ITERATION OVER DATASET
3. Using `tfds.load(..., shuffle_files=True)`. --> 61 seconds for 1 COMPLETE ITERATION OVER DATASET

These benchmarks suggest that the slowness of your pipeline is primarily caused by the positioning of your `shuffle` op and the incorrect buffer size you provide it.

**Comment by AdityaKane2001** [comment]
@kdonbekci 
Thanks for bringing this to notice! I had completely missed that 😅

**Comment by kdonbekci** [comment]
No problemo!

**Comment by braindotai** [comment]
Thanks a lot @AdityaKane2001 and @kdonbekci,

I tested out multiple ways of applying `.map` functions, though in all tests the results were better/worse by a margin of not more than 5-8 seconds.

# **Test1**:
On printing function name inside each function decorated by `tf.function`, the names for each function were printed just once. This implies tracing of function is happening just once. So no known performance regression is there due to `@tf.function`.
I also tried removing the decorator `tf.function` from all functions but found no noticable difference.

# **Test2**
On changing the different location of `.shuffle` with buffersize set to `train_data.cardinality()`, there was noticeable difference of 5-8 seconds based on different locations.

# **Major Difference Test**
**Here I completely removed the `.shuffle`, and voila, now we are down to ~6.5 seconds from minutes.**

So it's just the shuffling that's blocking the main process.

I'm guessing the reason why shuffling, in this case is so slow, is because usually the shuffling of file paths is done, instead of image files stored in memory.

Now in my use case, since I'm already applying a lot of random augmentations, so I don't think shuffling the data is that much required in the first place. So technically my issue is solved.

But still, any further clarifications of questions below would be really appreciated.

- Why shuffling is slow here exactly?
- And what an ideal approach would be for shuffling the image files stored in memory?

Thanks for your support.

Here's [the notebook](https://colab.research.google.com/gist/braindotai/3b917caf1d9f72569e5af1162d77ca42/untitled561.ipynb#scrollTo=o3BhLI-_2uI4) for the final test.

**Comment by kdonbekci** [comment]
Great to hear your input pipeline is matching your performance needs. As a compromise, I'd suggest you use the `shuffle_files` option of `tfds.load` and select a relatively low `shuffle_buffer_size` for an early `tf.data` `.shuffle` operation in your pipeline (It should really be the first thing). 

Your intuition is right that the `shuffle` operation is bottlenecking your pipeline. 
The reasons for this is a combination of:
1. Placement of your `.shuffle` at the end means that all the augmentations + batching need to happen before shuffling. Plus your unnecessarily high buffer size meant A TON of loading&preprocessing (although not quite sure on this anymore for what happens when `shuffle_buffer_size >> .cardinality()`) 
2. You are benchmarking the first example rather than a full iteration over the dataset. To understand the effect of this, let's say producing 1 batch with your augmentations takes `x` seconds and you want to shuffle with buffer size `b`. Now, since you've composed your pipeline to shuffle after the augmentations+batching is done, `tf.data` needs to load&augment `b` many batches so the time to come up with the first batch will take `b * x` seconds. Now, for the second iteration, `tf.data` needs to load&preprocess only a single batch to fill its shuffle buffer to the desired size. Thus the second batch will take `x` seconds. 
3. Google Colab machines have network attached storage devices which make loading files into RAM much slower than say reading it from your local SSD. With each `.tfrecord` file being ~100MB and the need to load all of them for a full shuffle, this could alone be the culprit. Like you say, for that reason, shuffling of file paths is preferred, which is what `tfds.load(shuffle_files=True)` does. `tf.data`'s `.interleave` (https://www.tensorflow.org/api_docs/python/tf/data/Dataset#interleave)  could help here, although complexifying your application a bit.

**Comment by braindotai** [comment]
@kdonbekci 
Thanks mate, really appreciate such detailed explanations.

Just want to highlight one thing mentioned by @AdityaKane2001, about the varying shape input to the functions decorated with `tf.function`, which is my case since many images are of different sizes. Based on official documentation, we can pass info in the `tf.function` containing input signature as:

```python
@tf.function(input_signature = (tf.TensorSpec(shape = [None, None, 3], dtype = tf.uint8)))
def function():
   ...
```
Since TensorFlow matches tensors based on their shape, using a `None` dimension as a wildcard will allow Functions to reuse traces for variably-sized input.

Alright guys, thanks again for the support :)
